pkg/kv/kvserver: BenchmarkBumpSideTransportClosed failed

pkg/kv/kvserver.BenchmarkBumpSideTransportClosed [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachprodMicrobenchmarkWeeklyWip/20289400?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachprodMicrobenchmarkWeeklyWip/20289400?buildTab=artifacts#./artifacts/microbench/experiment) on master @ [ca0d4076c381609d6487040141c87e36a68b211f](https://github.com/cockroachdb/cockroach/commits/ca0d4076c381609d6487040141c87e36a68b211f):


```
goos: linux
goarch: amd64
cpu: Intel(R) Xeon(R) CPU @ 2.60GHz
BenchmarkBumpSideTransportClosed
    test_log_scope.go:171: test logs captured to: /tmp/logBenchmarkBumpSideTransportClosed1507158493
    test_log_scope.go:82: use -show-logs to present logs inline
    replica_closedts_test.go:736: -- test log scope end --
    test_log_scope.go:171: test logs captured to: /tmp/logBenchmarkBumpSideTransportClosed354871636
    test_log_scope.go:82: use -show-logs to present logs inline
    replica_closedts_test.go:736: -- test log scope end --
    test_log_scope.go:171: test logs captured to: /tmp/logBenchmarkBumpSideTransportClosed1630114694
    test_log_scope.go:82: use -show-logs to present logs inline
    replica_closedts_test.go:736: -- test log scope end --
    test_log_scope.go:171: test logs captured to: /tmp/logBenchmarkBumpSideTransportClosed4113031375
    test_log_scope.go:82: use -show-logs to present logs inline
    replica_closedts_test.go:736: -- test log scope end --
    test_log_scope.go:171: test logs captured to: /tmp/logBenchmarkBumpSideTransportClosed3499275865
    test_log_scope.go:82: use -show-logs to present logs inline
    replica_closedts_test.go:733: BumpSideTransportClosed unexpectedly failed
    panic.go:636: -- test log scope end --
--- FAIL: BenchmarkBumpSideTransportClosed-32
FAIL

I250803 23:11:05.788308 1 (gostd) rand.go:268  [-] 1  random seed: 1809778845682530433
test logs left over in: /tmp/logBenchmarkBumpSideTransportClosed3499275865
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*BenchmarkBumpSideTransportClosed.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-53196

--- a/pkg/kv/kvserver/replica_closedts_test.go
+++ b/pkg/kv/kvserver/replica_closedts_test.go
@@ -694,6 +694,7 @@ func TestRejectedLeaseDoesntDictateClosedTimestamp(t *testing.T) {
 // we eliminate some memory copying and pass the node liveness record in to the
 // function so that we only have to grab it once instead of on each call to
 // BumpSideTransportClosed, we should be able to reach our target latency.
+// TODO(pav-kv): in 2025, this shows about 180-200ns.
 func BenchmarkBumpSideTransportClosed(b *testing.B) {
 	defer leaktest.AfterTest(b)()
 	defer log.Scope(b).Close(b)
@@ -720,6 +721,22 @@ func BenchmarkBumpSideTransportClosed(b *testing.B) {
 	manual.Pause()
 	now := s.Clock().NowAsClockTimestamp()
 	targets := map[ctpb.RangeClosedTimestampPolicy]hlc.Timestamp{}
+
+	// Wait for the scratch range's lease upgrade from expiration-based lease. The
+	// closed timestamp bumping requires for there to be no in-flight proposals,
+	// so a racing lease upgrade can cause a test failure like in #151295.
+	//
+	// TODO(pav-kv): share this code with TestCluster.MaybeWaitForLeaseUpgrade().
+	testutils.SucceedsSoon(b, func() error {
+		lease, _, err := s.GetRangeLease(ctx, key, roachpb.QueryLocalNodeOnly)
+		require.NoError(b, err)
+		typ := lease.Current().Type()
+		if typ == roachpb.LeaseExpiration {
+			return errors.Errorf("still on expiration-based lease")
+		}
+		b.Logf("lease is now of type: %s", typ)
+		return nil
+	})
 
 	b.ResetTimer()
 	for i := 0; i < b.N; i++ {

--- a/pkg/kv/kvserver/replica_closedts_test.go
+++ b/pkg/kv/kvserver/replica_closedts_test.go
@@ -730,7 +730,7 @@ func BenchmarkBumpSideTransportClosed(b *testing.B) {
 		// Perform the call.
 		res := r.BumpSideTransportClosed(ctx, now, targets)
 		if !res.OK {
-			b.Fatal("BumpSideTransportClosed unexpectedly failed")
+			b.Fatalf("BumpSideTransportClosed unexpectedly failed: %+v", res)
 		}
 	}
 }

